feat: improved grammar; less (no?) ambiguities;

alacambre-yseop opened the pull request:
Should parse the majority of YML files completely, except text granules (content is skipped).
Assumes that devs use a semi-colon at the end of most of the instructions.
→ The extension now displays warnings for parsing problems.

mminot-yseop commented on server/src/YmlToBdl.g4:
```diff
@@ -218,6 +254,11 @@ fieldValue : field
 
 //Functions
 
+function: (METHOD|FUNCTION|TEXT_METHOD|TEXT_FUNCTION) ymlId ( argsBlock  |  OPEN_PAR argumentList CLOSE_PAR ) localBlock? staticBlock? memberOption=field* SEMICOLON;
```
```suggestion
function: (METHOD|FUNCTION|TEXT_METHOD|TEXT_FUNCTION) ymlId (argsBlock|OPEN_PAR argumentList CLOSE_PAR) localBlock? staticBlock? memberOption=field* SEMICOLON;
```

**OR**

```suggestion
function: (METHOD | FUNCTION | TEXT_METHOD | TEXT_FUNCTION) ymlId (argsBlock | OPEN_PAR argumentList CLOSE_PAR) localBlock? staticBlock? memberOption=field* SEMICOLON;
```

?

alacambre-yseop commented on server/src/YmlToBdl.g4:
```diff
@@ -113,49 +127,57 @@ WS : [ \r\t\n]+ -> channel(HIDDEN);
 fragment MULTILINE_COMMENT_START: '/*';
 fragment MULTILINE_COMMENT_END: '*/';
 
+PREPROCESSING: '@' ~[\n\r]* -> channel(HIDDEN);
```
I don't know their structure at all, so I prefer to ignore them for now.

alacambre-yseop commented on server/src/YmlToBdl.g4:
```diff
@@ -113,49 +127,57 @@ WS : [ \r\t\n]+ -> channel(HIDDEN);
 
 // Colons aren't interpreted by YE in a particular way.
-ymlId : ID (COLON? COLON ID)*;
+
```
I changed the comments here and there.

alacambre-yseop commented:
@mminot-yseop I think this version of the grammar seems stable enough to do new devs upon it and is, therefore, ready for review.
I also added some documentation and updated the `CHANGELOG` file.

mminot-yseop commented on client/README.md:
```diff
@@ -6,56 +6,58 @@ This extension provides support for Yseop Markup Language (YML).
 
+### Parsing error warnings
+
+![Example image for parsing warning](https://raw.githubusercontent.com/yseop/vscode-yseopml/master/client/images/parsingProblem.png "Parsing warnings in the “Problem”'s and editor's view.")
```
```suggestion
![Example image for parsing warning](https://raw.githubusercontent.com/yseop/vscode-yseopml/master/client/images/parsingProblem.png "Parsing warnings in the “Problems” and editor views.")
```

mminot-yseop commented on client/README.md:
```diff
@@ -6,56 +6,58 @@ This extension provides support for Yseop Markup Language (YML).
+- `textu`: Textualization attribute
+- `\thirdAction`: Verb using a third person subject
+- `\value`: Display a value
+- `\verb`: Verb
```
Why change the bullets here and not in the script whose purpose is to generate the list?

alacambre-yseop replied:
Automatic formatting, sorry; this was an unexpected change.

mminot-yseop commented:
Haha, honestly, I was curious to see how you would handle that hyphen affair.

```
$ printf '- `%s`: %s\n' foo bar
bash: printf: - : invalid option
printf: usage: printf [-v var] format [arguments]
```

Hehe. Remember that `--` trick to say “no options from here onwards, only positional arguments”? Now's the time to use it. :laughing:

alacambre-yseop commented:
@mminot-yseop It didn't failed on my computer, that's why I pushed it.

alacambre-yseop commented:
Multiple misclicks, sorry.

mminot-yseop commented:
> @mminot-yseop It didn't failed on my computer, that's why I pushed it.

That's rather strange.
Hum… It seems `xargs` somehow makes that work. I guess it's safer to add the `--` anyway to be more implementation-independent or something.

alacambre-yseop commented:
@mminot-yseop Done.

alacambre-yseop commented:
@mminot-yseop It's OK for me in its current state.

mminot-yseop commented:
I'll try opening random files with it, then, I guess.

mminot-yseop commented:
Now that I think about it, it's kinda weird to make references to Enhancement Fridays in branch names in a public project!

mminot-yseop commented:
Hum, so since you ignore `@foo` lines, there is no check that, for example, `@if` has a corresponding `@endif`. (My VSC opened on an old `project.kao` file that was full of these.)

mminot-yseop commented:
```
```

alacambre-yseop commented:
@mminot-yseop `project.kao`-like files and preprocessing instructions aren't handled yet but will be in the future.
I added an option in case the user wants to disable the warnings.
